File: packages/sdk-python/src/melaya/credentials.py
"""Credentials API — store, retrieve, test, and delete secrets and service connections.

Maps to ``/api/v1/private/credentials/*``. Credentials are envelope-encrypted at rest.
Use ``melaya.connectors`` for project-scoped connector credentials.

Example
-------
>>> from melaya import Melaya
>>> m = Melaya(api_key="mk_...")
>>> m.credentials.set("openai", value="sk-...", label="OpenAI prod key")
>>> result = m.credentials.test("openai")
>>> models = m.credentials.list_models(provider="openai")
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from .platform_types import JsonDict


class CredentialsAPI:
    def __init__(self, request: Any) -> None:
        self._request = request

    def list(self) -> List[JsonDict]:
        """List all stored credentials (services, OAuth connections, env handles)."""
        return self._request("GET", "/api/v1/private/credentials")

    def connected_services(self) -> List[JsonDict]:
        """List connected third-party services for the caller."""
        return self._request("GET", "/api/v1/private/credentials/services")

    def get_operator_profile(self) -> JsonDict:
        """Get operator profile (persona config for agent context)."""
        return self._request("GET", "/api/v1/private/credentials/operator-profile")

    def get_luma_schema(self, *, event_id: Optional[str] = None) -> JsonDict:
        """Get the Luma event registration form schema."""
        params: Dict[str, Any] = {}
        if event_id is not None:
            params["eventId"] = event_id
        return self._request("GET", "/api/v1/private/credentials/luma/schema", params=params)

    def set_operator_profile(self, profile: Dict[str, Any]) -> JsonDict:
        """Save operator profile."""
        return self._request("PUT", "/api/v1/private/credentials/operator-profile", json=profile)

    def get(self, service: str, *, key: Optional[str] = None) -> JsonDict:
        """Get a stored credential value by service; pass key to retrieve a specific named key."""
        params: Dict[str, Any] = {}
        if key is not None:
            params["key"] = key
        return self._request("GET", f"/api/v1/private/credentials/{service}", params=params if params else None)

    def set(self, service: str, *, value: str, key: Optional[str] = None, label: Optional[str] = None) -> JsonDict:
        """Store or update a credential (envelope-encrypted at rest)."""
        body: Dict[str, Any] = {"value": value}
        if key is not None:
            body["key"] = key
        if label is not None:
            body["label"] = label
        return self._request("PUT", f"/api/v1/private/credentials/{service}", json=body)

    def delete(self, service: str) -> JsonDict:
        """Delete a stored credential by service name."""
        return self._request("DELETE", f"/api/v1/private/credentials/{service}")

    def rag_ingest_start(self, **kwargs: Any) -> JsonDict:
        """Start a RAG document ingestion job."""
        return self._request("POST", "/api/v1/private/rag/ingest", json=kwargs)

    def rag_ingest_status(self, session_id: str) -> JsonDict:
        """Poll RAG ingestion job status."""
        return self._request("GET", f"/api/v1/private/rag/ingest/{session_id}")

    def rag_retrieve_start(self, **kwargs: Any) -> JsonDict:
        """Start a RAG retrieval query."""
        return self._request("POST", "/api/v1/private/rag/retrieve", json=kwargs)

    def rag_retrieve_status(self, session_id: str) -> JsonDict:
        """Poll RAG retrieval result."""
        return self._request("GET", f"/api/v1/private/rag/retrieve/{session_id}")

    def pick_folder_start(self) -> JsonDict:
        """Initiate native folder picker for file ingestion."""
        return self._request("POST", "/api/v1/private/rag/pick-folder")

    def pick_folder_status(self, session_id: str) -> JsonDict:
        """Poll folder picker result."""
        return self._request("GET", f"/api/v1/private/rag/pick-folder/{session_id}")

    def linkedin_connect_start(self) -> JsonDict:
        """Start LinkedIn OAuth flow."""
        return self._request("POST", "/api/v1/private/credentials/linkedin/connect")

    def linkedin_connect_cancel(self) -> JsonDict:
        """Cancel an in-progress LinkedIn OAuth flow."""
        return self._request("DELETE", "/api/v1/private/credentials/linkedin/connect")

    def linkedin_connect_status(self) -> JsonDict:
        """Poll LinkedIn OAuth connection status."""
        return self._request("GET", "/api/v1/private/credentials/linkedin/connect/status")

    def luma_connect_start(self) -> JsonDict:
        """Start Luma OAuth flow."""
        return self._request("POST", "/api/v1/private/credentials/luma/connect")

    def luma_connect_status(self) -> JsonDict:
        """Poll Luma OAuth connection status."""
        return self._request("GET", "/api/v1/private/credentials/luma/connect/status")

    def luma_connect_cancel(self) -> JsonDict:
        """Cancel an in-progress Luma OAuth flow."""
        return self._request("DELETE", "/api/v1/private/credentials/luma/connect")

    def test(self, service: str) -> JsonDict:
        """Test a stored credential (e.g. validate API key against the exchange)."""
        return self._request("POST", f"/api/v1/private/credentials/{service}/test")

    def melaya_accounts(self) -> List[JsonDict]:
        """List Melaya sub-accounts available to the caller."""
        return self._request("GET", "/api/v1/private/credentials/melaya-accounts")

    def google_oauth_start(self) -> JsonDict:
        """Start Google OAuth flow for credential storage."""
        return self._request("POST", "/api/v1/private/credentials/google/oauth")

    def cli_auth_start(self) -> JsonDict:
        """Start CLI authentication flow (device-code style)."""
        return self._request("POST", "/api/v1/private/credentials/cli-auth")

    def notebooklm_login(self, *, username: Optional[str] = None, password: Optional[str] = None, **kwargs: Any) -> JsonDict:
        """Store NotebookLM credentials."""
        body: Dict[str, Any] = {**kwargs}
        if username is not None:
            body["username"] = username
        if password is not None:
            body["password"] = password
        return self._request("POST", "/api/v1/private/credentials/notebooklm/login", json=body)

    def notebooklm_status(self) -> JsonDict:
        """Check NotebookLM connection status."""
        return self._request("GET", "/api/v1/private/credentials/notebooklm/status")

    def telegram_auth_start(self, *, phone: str) -> JsonDict:
        """Start Telegram user auth (phone number step)."""
        return self._request("POST", "/api/v1/private/credentials/telegram/auth", json={"phone": phone})

    def telegram_auth_code(self, *, code: str) -> JsonDict:
        """Submit Telegram SMS verification code."""
        return self._request("POST", "/api/v1/private/credentials/telegram/auth/code", json={"code": code})

    def telegram_auth_2fa(self, *, password: str) -> JsonDict:
        """Submit Telegram 2FA password."""
        return self._request("POST", "/api/v1/private/credentials/telegram/auth/2fa", json={"password": password})

    def list_models(self, *, provider: Optional[str] = None, capability: Optional[str] = None) -> List[JsonDict]:
        """List available AI models across all configured providers.

        Collapses 19+ provider fan-out into a parameterized query.
        """
        params: Dict[str, Any] = {}
        if provider is not None:
            params["provider"] = provider
        if capability is not None:
            params["capability"] = capability
        return self._request("GET", "/api/v1/private/credentials/models", params=params if params else None)

File: packages/sdk-ruby/lib/melaya/credentials.rb
# frozen_string_literal: true

module Melaya
  # Credentials API — store, retrieve, test, and delete secrets and
  # third-party service connections at user scope.
  #
  # Credentials are envelope-encrypted at rest. Use +Melaya::ConnectorsAPI+
  # for project-scoped connector credentials.
  #
  # Maps to /api/v1/private/credentials/*.
  #
  # @example
  #   melaya.credentials.set("openai", value: "sk-...")
  #   melaya.credentials.test("openai")
  #   melaya.credentials.delete("openai")
  class CredentialsAPI
    def initialize(http)
      @http = http
    end

    # GET /api/v1/private/credentials
    # List all stored credentials (services, OAuth connections, env handles).
    def list
      @http.get("/api/v1/private/credentials")
    end

    # GET /api/v1/private/credentials/services
    # List connected third-party services for the caller.
    def connected_services
      @http.get("/api/v1/private/credentials/services")
    end

    # GET /api/v1/private/credentials/:service
    # Get a stored credential value by service name.
    # @param service [String]
    # @param key [String, nil] optional named key within the service
    def get(service, key: nil)
      params = key ? { "key" => key } : {}
      @http.get("/api/v1/private/credentials/#{enc(service)}", params)
    end

    # PUT /api/v1/private/credentials/:service
    # Store or update a credential (envelope-encrypted at rest).
    # @param service [String]
    # @param value [String] the secret value
    # @param key [String, nil]
    # @param label [String, nil]
    def set(service, value:, key: nil, label: nil)
      body = compact("value" => value, "key" => key, "label" => label)
      @http.put("/api/v1/private/credentials/#{enc(service)}", body)
    end

    # DELETE /api/v1/private/credentials/:service
    # Delete a stored credential by service name.
    # @param service [String]
    def delete(service)
      @http.delete("/api/v1/private/credentials/#{enc(service)}")
    end

    # POST /api/v1/private/credentials/:service/test
    # Test a stored credential (e.g. validate API key against its service).
    # @param service [String]
    def test(service)
      @http.post("/api/v1/private/credentials/#{enc(service)}/test")
    end

    # ── Operator profile ───────────────────────────────────────────────────────

    # GET /api/v1/private/credentials/operator-profile
    # Get operator profile (persona config for agent context).
    def get_operator_profile
      @http.get("/api/v1/private/credentials/operator-profile")
    end

    # PUT /api/v1/private/credentials/operator-profile
    # Save operator profile.
    # @param profile [Hash]
    def set_operator_profile(profile)
      @http.put("/api/v1/private/credentials/operator-profile", profile)
    end

    # ── AI models ──────────────────────────────────────────────────────────────

    # GET /api/v1/private/credentials/models
    # List available AI models across all configured providers.
    # Collapses 19+ provider fan-out into a parameterized query.
    # @param provider [String, nil]
    # @param capability [String, nil]
    def list_models(provider: nil, capability: nil)
      @http.get("/api/v1/private/credentials/models",
        compact("provider" => provider, "capability" => capability))
    end

    # ── Melaya sub-accounts ────────────────────────────────────────────────────

    # GET /api/v1/private/credentials/melaya-accounts
    # List Melaya sub-accounts available to the caller.
    def melaya_accounts
      @http.get("/api/v1/private/credentials/melaya-accounts")
    end

    # ── RAG ────────────────────────────────────────────────────────────────────

    # POST /api/v1/private/rag/ingest
    # Start a RAG document ingestion job.
    # @param body [Hash]
    def rag_ingest_start(body = {})
      @http.post("/api/v1/private/rag/ingest", body)
    end

    # GET /api/v1/private/rag/ingest/:sessionId
    # Poll RAG ingestion job status.
    # @param session_id [String]
    def rag_ingest_status(session_id)
      @http.get("/api/v1/private/rag/ingest/#{enc(session_id)}")
    end

    # POST /api/v1/private/rag/retrieve
    # Start a RAG retrieval query.
    # @param body [Hash]
    def rag_retrieve_start(body = {})
      @http.post("/api/v1/private/rag/retrieve", body)
    end

    # GET /api/v1/private/rag/retrieve/:sessionId
    # Poll RAG retrieval result.
    # @param session_id [String]
    def rag_retrieve_status(session_id)
      @http.get("/api/v1/private/rag/retrieve/#{enc(session_id)}")
    end

    # ── Folder picker ──────────────────────────────────────────────────────────

    # POST /api/v1/private/rag/pick-folder
    # Initiate native folder picker for file ingestion.
    # @param body [Hash]
    def pick_folder_start(body = {})
      @http.post("/api/v1/private/rag/pick-folder", body)
    end

    # GET /api/v1/private/rag/pick-folder/:sessionId
    # Poll folder picker result.
    # @param session_id [String]
    def pick_folder_status(session_id)
      @http.get("/api/v1/private/rag/pick-folder/#{enc(session_id)}")
    end

    # ── LinkedIn OAuth ─────────────────────────────────────────────────────────

    # POST /api/v1/private/credentials/linkedin/connect
    # Start LinkedIn OAuth flow.
    def linkedin_connect_start(body = {})
      @http.post("/api/v1/private/credentials/linkedin/connect", body)
    end

    # DELETE /api/v1/private/credentials/linkedin/connect
    # Cancel an in-progress LinkedIn OAuth flow.
    def linkedin_connect_cancel
      @http.delete("/api/v1/private/credentials/linkedin/connect")
    end

    # GET /api/v1/private/credentials/linkedin/connect/status
    # Poll LinkedIn OAuth connection status.
    def linkedin_connect_status
      @http.get("/api/v1/private/credentials/linkedin/connect/status")
    end

    # ── Luma OAuth ────────────────────────────────────────────────────────────

    # POST /api/v1/private/credentials/luma/connect
    # Start Luma OAuth flow.
    def luma_connect_start(body = {})
      @http.post("/api/v1/private/credentials/luma/connect", body)
    end

    # GET /api/v1/private/credentials/luma/connect/status
    # Poll Luma OAuth connection status.
    def luma_connect_status
      @http.get("/api/v1/private/credentials/luma/connect/status")
    end

    # DELETE /api/v1/private/credentials/luma/connect
    # Cancel an in-progress Luma OAuth flow.
    def luma_connect_cancel
      @http.delete("/api/v1/private/credentials/luma/connect")
    end

    # GET /api/v1/private/credentials/luma/schema
    # Get the Luma event registration form schema.
    # @param params [Hash]
    def luma_schema(params = {})
      @http.get("/api/v1/private/credentials/luma/schema", params)
    end

    # ── Google OAuth ───────────────────────────────────────────────────────────

    # POST /api/v1/private/credentials/google/oauth
    # Start Google OAuth flow for credential storage.
    def google_oauth_start(body = {})
      @http.post("/api/v1/private/credentials/google/oauth", body)
    end

    # ── CLI auth ───────────────────────────────────────────────────────────────

    # POST /api/v1/private/credentials/cli-auth
    # Start CLI authentication flow (device-code style).
    def cli_auth_start(body = {})
      @http.post("/api/v1/private/credentials/cli-auth", body)
    end

    # ── NotebookLM ─────────────────────────────────────────────────────────────

    # POST /api/v1/private/credentials/notebooklm/login
    # Store NotebookLM credentials.
    def notebooklm_login(body = {})
      @http.post("/api/v1/private/credentials/notebooklm/login", body)
    end

    # GET /api/v1/private/credentials/notebooklm/status
    # Check NotebookLM connection status.
    def notebooklm_status
      @http.get("/api/v1/private/credentials/notebooklm/status")
    end

    # ── Telegram auth ──────────────────────────────────────────────────────────

    # POST /api/v1/private/credentials/telegram/auth
    # Start Telegram user auth (phone number step).
    def telegram_auth_start(body = {})
      @http.post("/api/v1/private/credentials/telegram/auth", body)
    end

    # POST /api/v1/private/credentials/telegram/auth/code
    # Submit Telegram SMS verification code.
    def telegram_auth_code(body = {})
      @http.post("/api/v1/private/credentials/telegram/auth/code", body)
    end

    # POST /api/v1/private/credentials/telegram/auth/2fa
    # Submit Telegram 2FA password.
    def telegram_auth_2fa(body = {})
      @http.post("/api/v1/private/credentials/telegram/auth/2fa", body)
    end

    private

    def enc(s)
      URI.encode_www_form_component(s.to_s)
    end

    def compact(hash)
      hash.reject { |_, v| v.nil? }
    end
  end
end
